Add nan-comparison checker for NaN values

orSolocate opened the pull request:
<!--

Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to pylint in this document:
https://github.com/PyCQA/pylint/blob/master/doc/development_guide/contribute.rst#repository
-->

# Steps
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x]  Add a ChangeLog entry.
- [x]  If it's a new feature or an important bug fix, add a What's New entry in doc/whatsnew/<current release.rst>.
- [x]  Write a good description on what the PR does.

# Description
 #### Add `nan-comparison` message for comparison of NaN values (e.g. numpy.NaN or float('nan')).
`NaN` (Not A Number) is a special entity that does not equal to any other entity, including itself.
Therefore a comparison `==` between any object and `NaN` will yield `False`. The correct way to check whether a variable is NaN would be using the `is` operator.

Implementation of this checker uses the infrastructure of `singleton-comparison` checker and is under the 
`base checker -> comparison checker`,
functional test and unittest added and similar to the `singleton-comparison checker`.

Example:
`x==numpy.NaN` statement would always be False, and should be `x is numpy.NaN`

# Type of Changes

IsDone | Type
------------ | -------------
|| 🐛 Bug fix
✓|✨ New feature
||🔨  Refactoring
||📜  Docs

# Related Issue
None

hippo91 commented on pylint/checkers/base.py:
```diff
@@ -2399,6 +2405,27 @@ def _is_singleton_const(node) -> bool:
+    ):
+        nan_values = ("float('nan')", "numpy.NaN", "np.NaN")
+
+        def _is_NaN(node) -> bool:
+            return node.as_string() in nan_values
```
This way of checking the Nan value is not robust. What if, for example, `numpy` is imported as `nmp`?
Instead we should ask astroid to infer the value of the node.
For example in case of `float('nan')` we can use astroid this way:
```
 if isinstance(node, astroid.Call) and node.args[0].value == 'nan' and node.inferred()[0].pytype() == 'builtins.float':
     return True
```

orSolocate replied:
Okay. will be fixed.

hippo91 reviewed (changes requested):
@orSolocate thanks for this very clear and well tested PR. I left a bunch of comments/remarks that i let you discover.

hippo91 commented on tests/functional/n/nan_comparison.py:
```diff
@@ -0,0 +1,21 @@
+x = 42
+a = x is numpy.NaN
+b = x == numpy.NaN  # [nan-comparison]
+c = x == float('nan')  # [nan-comparison]
```
I am bit confused here.IMHO the only clean way to test if a float value is nan is to use the `isnan` function of the `math` module. For example:
```
x = float('nan')

x == float('nan')
False

x is float('nan')
False

import math

math.isnan(x)
True

x is math.nan
False
```

orSolocate replied:
Okay i take your advice. when the operator is one of 'is' 'is not' '==' '!=' i would suggest to change it to math.isnan(<>)

orSolocate commented on pylint/checkers/base.py:
```diff
@@ -2339,6 +2339,12 @@ class ComparisonChecker(_BasicChecker):
+        "W0177": (
+            "Comparison %s should be %s",
+            "nan-comparison",
+            "Used when an expression is compared to NaN value"
```
Okay I agree.

orSolocate commented on pylint/checkers/base.py:
```diff
@@ -2399,6 +2405,27 @@ def _is_singleton_const(node) -> bool:
+            return
+
+        nan_comparison_example = {False: "'{} is {}'", True: "'{} is not {}'"}
+        suggestion = nan_comparison_example[checking_for_absence].format(
```
@hippo91  That's the way they do it in `singleton-comparison` under the same file.
But I can change it to something like:

if <!cond>: is_msg= ' is '
else: is_msg=' is not '

and then in the message itself concatenate it.

hippo91 replied:
@orSolocate do not bother with that. If it has been accepted in another part of the same file there is no reason to change.
Apologize.

orSolocate commented on tests/checkers/unittest_base.py:
```diff
@@ -444,6 +444,29 @@ def test_comparison(self):
+            args=("'foo != numpy.NaN'", "'foo is not numpy.NaN'"),
+        )
+        with self.assertAddsMessages(message):
+            self.walk(node)
```
Ok no problem

hippo91 replied:
Thanks!

hippo91 commented on pylint/checkers/base.py:
```diff
@@ -2408,18 +2408,50 @@ def _is_singleton_const(node) -> bool:
 
         def _is_NaN(node) -> bool:
-            return node.as_string() in nan_values
+            if _is_float_nan(node):
+                return True
+            if _is_numpy_NaN(node):
+                return True
```
Why not 
```
return _is_float_nan(node) or _is_float_Nan(node)
```?

orSolocate replied:
sure. will change that.

hippo91 commented on pylint/checkers/base.py:
```diff
@@ -2408,18 +2408,50 @@ def _is_singleton_const(node) -> bool:
+            except (AttributeError, IndexError):
+                return False
+
+        def _is_numpy_NaN(node):
```
Would it be possible to have uniform case for `nan` word?

orSolocate replied:
like `_is_nan_word` method that covers all the cases for `nan`?
I could do that.

hippo91 replied:
I am not sure to understand your answer. I just wanted to say that it could be nice to have the word `nan` (in lowercase) or `NaN` in all method names.

orSolocate replied:
Now I understand what you meant, will fix it

hippo91 reviewed (approved):
@orSolocate  and @Or Bahari thanks for the modifications. Once the CI will be ok we could merge this into master. I just left two comments that should be addressed before merging.

orSolocate commented:
@hippo91  I found out that the CI does not pass because of this line:
`inferred_module = next(node.expr.infer())` in my numpy detection method:

    def _is_numpy_nan(node):
            if isinstance(node, astroid.Attribute) and node.attrname == "NaN":
                if isinstance(node.expr, astroid.Name):
                    inferred_module = next(node.expr.infer())
                    if isinstance(inferred_module, astroid.Module) and inferred_module.name == "numpy":
                        return True
            return False

The test_functional.py and unittest pass with Pytest so I don't understand where the problem is coming from when running the tox tests. I tried to change it to:
`next(astroid.inference.infer_name(node.expr))`
or to change the infer function to: `_infer` but it doesn't help. Is it a known issue?

I think the best thing to do would be to upload the newly uplaoded version of `is_numpy_nan` and open a bug/feature request to support all inferred names of the module. Practically in my opinion there is a very low chance that someone would use a different name rather than 'np','numpy' or 'nmp' or to use this names with a different class without importing the `numpy` package with an attribute named 'NaN'... I feel the checker as it is now adds value to Pylint.

hippo91 commented:
@orSolocate it seems that the CI of your last commit is ok. Thus i think your previous commit is no longer up to date.

hippo91 reviewed (commented):
@orSolocate i just have a couple of comments. I'am also investigating the problem of your last comment.

hippo91 commented:
@orSolocate the problem of CI failure with your previous implementation of the `_is_numpy_nan` seems to be due to the fact that `numpy` is not present in the tox environments.
Before creating an environment with a `numpy` dependency i would like to have @PCManticore opinion on this subject. Can we imagine the same testing of `numpy` components as the one present in `astroid`?

orSolocate commented on tests/functional/n/nan_comparison_check.txt:
```diff
@@ -0,0 +1,12 @@
+nan-comparison:6::Comparison 'x is numpy.NaN' should be 'math.isnan(x)'
+nan-comparison:7::Comparison 'x == numpy.NaN' should be 'math.isnan(x)'
```
why? to avoid importing the `math` package? in that case one could just do:
`from math import isnan`
I just don't trust as much the numpy implementation that could change in the future, as opposed to the math implementation which is just:

`def isnan(x):
    return x!=x`

@hippo91  what do you think?

hippo91 replied:
@orSolocate sorry i may have been unclear. What i wanted to say is that when people already use `numpy` such as in `x == numpy.NaN` then `x` is probably a `numpy` object and then IMHO it make more sense to recommend to use `numpy.isnan(x)`.
Of course when people are just dealing with regular float there is no reason to ask them using `numpy`.

orSolocate replied:
I decided not to make the fix. I would the `numpy` suggestion to be according to the name the user refers to `numpy` through. for example if the user uses `import numpy as np` i want to suggest to use `'np.isnan(x)'` . and I can't do that before you guys make your fix around the numpy library to include in your environment (the infer to numpy lib won't pass).
I decided to upload it as it is now (after rebase) and open a new feature to fix the test env, as well as to add this as an extended feature. sounds good? @hippo91

orSolocate commented on pylint/checkers/base.py:
```diff
@@ -2399,6 +2403,52 @@ def _is_singleton_const(node) -> bool:
+            try:
+                if isinstance(node, astroid.Call) and len(node.args) == 1:
+                    if (
+                        node.args[0].value == "nan"
```
yeah it is a good idea thanks for noticing it. I didn't know you could do float('NAN') with uppercase.. I could add a fix

orSolocate commented:
@hippo91  Okay, so I guess there is nothing I could do about it. I will just fix according to your suggestions. Do you want me to squash all the commits to one commit rebased to master?

hippo91 commented:
@orSolocate you are right, for the moment, there is nothing to do. We will wait an answer of @PCManticore.

orSolocate commented:
squashed all my commits and rebased to master. looks good? 
@hippo91 also  see  my comment below regarding your last suggestion. I will open a separate issue for this ('new feature').

orSolocate commented:
I have no idea why this regression test `regression_property_no_member_2641` fails, i didn't change anything related to it. running this specific function test on pytest `test_functional[regression_property_no_member_2641]` passed but tox --recreate fails. Hard to debug it. @hippo91  any idea what causes this regression?

hippo91 commented:
@orSolocate this test fails because of my last PR on astroid. I will fix it ASAP. Do not bother with it.

Pierre-Sassoulas commented:
Hi, the master branch is fixed we could merge. Could you rebase on the latest master @orSolocate, please ?

orSolocate commented:
@Pierre-Sassoulas  Great I see that it passes the CI now, thanks! You can merge it now.
Or

Pierre-Sassoulas commented on doc/whatsnew/2.6.rst:
```diff
@@ -14,6 +14,8 @@ Summary -- Release highlights
 New checkers
 ============
 
+* Add `nan-comparison` check for comparison of NaN values
```
Could you rebase and move that to `2.7.rst`, please ? Sory it's due to #4095 being merged.

orSolocate replied:
Okay sure

Pierre-Sassoulas commented on doc/whatsnew/2.6.rst:
```diff
@@ -14,6 +14,13 @@ Summary -- Release highlights
 New checkers
 ============
 
+* Add `nan-comparison` check for comparison of NaN values
+
+* Add support to ``ignored-argument-names`` in DocstringParameterChecker and
+  adds `useless-param-doc` and `useless-type-doc` messages.
+
+* Add `empty-comment` check for empty comments.
+
```
Thank you for merging master in this. I think you still need to move this text from 2.6.rst to 2.7.rst :)

Pierre-Sassoulas commented:
Thank you for this new check and for being persistent until it was merged ;) It will be released in `2.7.0` !
